// @aws-lambda-powertools/logger/lib/formatter/PowertoolLogFormatter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PowertoolLogFormatter = void 0;
const _1 = require(".");
/**
 * This class is used to transform a set of log key-value pairs
 * in the AWS Lambda Powertools' default structure log format.
 *
 * @class
 * @extends {LogFormatter}
 */
class PowertoolLogFormatter extends _1.LogFormatter {
    /**
     * It formats key-value pairs of log attributes.
     *
     * @param {UnformattedAttributes} attributes
     * @returns {PowertoolLog}
     */
    formatAttributes(attributes) {
        return {
            cold_start: attributes.lambdaContext?.coldStart,
            function_arn: attributes.lambdaContext?.invokedFunctionArn,
            function_memory_size: attributes.lambdaContext?.memoryLimitInMB,
            function_name: attributes.lambdaContext?.functionName,
            function_request_id: attributes.lambdaContext?.awsRequestId,
            level: attributes.logLevel,
            message: attributes.message,
            sampling_rate: attributes.sampleRateValue,
            service: attributes.serviceName,
            timestamp: this.formatTimestamp(attributes.timestamp),
            xray_trace_id: attributes.xRayTraceId,
        };
    }
}
exports.PowertoolLogFormatter = PowertoolLogFormatter;
